src/metabase/transforms/core.clj

```diff
@@ -162,54 +131,47 @@
-                             requires)]
+  (let [tables   (for [table (table/with-fields
+                               (db/select 'Table :db_id db-id :schema schema))]
+                   ;; temporarily until we merge the sync PR
+                   (assoc table :domain_entity (de/domain-entity-for-table table)))
+        bindings (into {}
+                       (map (fn [requirement]
+                              (let [t (get-in @domain-entity-specs [requirement :type])]
+                                [requirement
+                                 (filter #(-> % :domain_entity :type (isa? t)) tables)])))
+                       requires)]
     ;; If multiple tables match punt for now
-    (when (every? (comp #{1} count second) bindings)
+    (when (every? (comp #{1} count val) bindings)
       (m/map-vals (fn [[table]]
                     {:entity     table
-                     :dimensions (into {} (for [field (:fields table)]
-                                            [(-> field field-type name) (mbql-reference field)]))})
+                     :dimensions (m/map-vals de/mbql-reference
+                                             (get-in table [:domain_entity :dimensions]))})
                   bindings))))
 
-(defn- store-requirements!
-  [db-id requirements]
+(s/defn ^:private store-requirements!
+  [db-id :- su/IntGreaterThanZero, bindings :- Bindings]
   (qp.store/fetch-and-store-database! db-id)
   (qp.store/fetch-and-store-fields!
-   (mapcat (comp (partial map u/get-id) :fields :entity val) requirements)))
+   (mapcat (comp (partial map u/get-id) :fields :entity val) bindings)))
 
 (s/defn apply-transform!
   "Apply transform defined by transform spec `spec` to schema `schema` in database `db-id`."
   [db-id :- su/IntGreaterThanZero, schema :- (s/maybe s/Str), {:keys [steps provides] :as spec} :- TransformSpec]
   (materialize/fresh-collection-for-transform! spec)
-  (let [initial-bindings (satisfy-requirements db-id schema spec)]
+  (when-let [initial-bindings (satisfy-requirements db-id schema spec)]
     (driver/with-driver (-> db-id Database :engine)
       (qp.store/with-store
         (store-requirements! db-id initial-bindings)
         (let [bindings (reduce transform-step! initial-bindings (vals steps))]
-          (for [[result-step {required-dimensions :dimensions}] provides]
-            (do
-              (when (not-every? (get-in bindings [result-step :dimensions]) required-dimensions)
-                (throw (Exception. (str (tru "Resulting transform {0} does not conform to expectations.\nExpected: {1}\nGot: {2}"
-                                             result-step
-                                             required-dimensions
-                                             (->> result-step bindings :dimensions keys))))))
-              (-> result-step bindings :entity u/get-id))))))))
+          (for [domain-entity-name provides]
```

Comment:
since you're binding `*driver*` and the QP store you should use either `doall` or `vec` here to force the lazy seq generated by `for`, otherwise those bindings won't be in play when the body of the `for` gets evaluated:

e.g.

```clj
(def ^:dynamic *dynamic-var* :x)

(defn x []
  (binding [*dynamic-var* :y]
    (for [i (range 100)]
      *dynamic-var*)))

(drop 99 (x)) ;; -> ❌ 

```

Reply:
vs

```clj
(defn y []
  (binding [*dynamic-var* :y]
    (doall
     (for [i (range 100)]
       *dynamic-var*))))

(drop 99 (y)) ;-> [:y]
```

Reply:
bump

Reply:
There's no for in there. That code is completely refactored